lib/puppet/type.rb

```diff
@@ -2117,8 +2117,17 @@ def autorelation(rel_type, rel_catalog = nil)
     auto_rel = "eachauto#{rel_type}".to_sym
 
     self.class.send(auto_rel) { |type, block|
-      # Ignore any types we can't find, although that would be a bit odd.
-      next unless Puppet::Type.type(type)
+      if not Puppet::Type.type(type)
+        # We know it is not a type at this point, let's see if it exists as a defined type in the catalog
+        # We split and join on :: because defined types will likely have colon separators
+        # All defined types are of type :component
+        found_defines = rel_catalog.resources.select do |x|
+          x.type == :component && x.title_class == type.to_s.split('::').map(&:capitalize).join('::')
```

Comment:
Is that @reference in type/component.rb ? (Not sure where to search for it). In general I don't mind if the instances used have methods that return what kind of puppet "kind" they are via a predicate method. It is very confusing the way it is in terms of which ruby instances represent what in puppet terms. IIRC, there is a predicate for 'is builtin', but since 'not is_builtin' is not the same as 'definition?' there seems to me something missing.